File: palsplit.c
/* palsplit.c
 *
 * Head/tail split search for palindromic curve-values
 * p = 2n^2 + 2n + 1 = n^2 + (n+1)^2.
 *
 * A d-digit palindrome's high t digits are the reverse of its low t
 * digits.  Enumerating r = n mod 10^t therefore fixes BOTH ends of p
 * at once: the low t digits directly (p = f(r) mod 10^t), and the
 * high t digits as their reverse.  The high digits in turn pin n to a
 * narrow band via n ~ sqrt(p/2).  Intersecting "n in band" with
 * "n = r (mod 10^t)" leaves ~10^(d/2 - 2t) candidates per residue over
 * 10^t residues; balanced at t ~ d/4 this is O(10^(d/4)), against
 * O(10^(d/2)) for straight n-enumeration (hunt.c, palhunt_gmp).
 *
 * This works because a palindrome's head is DETERMINED by its tail.
 * It does NOT carry over to the emirp search, where the head and tail
 * of p are independent choices -- see docs/palindrome_split_search.md.
 *
 * Correctness does not depend on t: p_i = p_(d-1-i) holds for every
 * t <= d, so every t returns the same set.  t only sets the speed.
 * Running two different t is therefore a free cross-check, and is
 * MANDATORY before believing any result past the d=27 frontier.
 *
 * Build:
 *   gcc -O3 -march=znver2 -mtune=znver2 -std=c99 -Wall -Wextra \
 *       -fopenmp -o palsplit palsplit.c -lgmp -lm
 *
 * Usage:
 *   ./palsplit <d> [t] [--keep5]
 *
 *     d        decimal digit length.  Odd d only in practice: curve
 *              values are never 0 mod 11, and every even-digit
 *              palindrome is divisible by 11, so no even d has a
 *              curve palindrome of any kind.
 *     t        split width (default d/4, the balance point).
 *     --keep5  keep palindromes divisible by 5.  These cannot be
 *              prime, but the pals_d*.txt regression sets list some
 *              of them, so the regression needs this flag.
 *
 * Example:
 *   $ ./palsplit 27 7 --keep5
 *   d=27 t=7 keep5=1 outer=10^7
 *   108491007414868414700194801  n=7365154696775   COMPOSITE
 *   318216440234333432044612813  n=12613810689762  COMPOSITE
 *   318288756988131889657882813  n=12615243893562  COMPOSITE
 *   found=3
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <math.h>
#include <float.h>
#include <omp.h>
#include <gmp.h>

#define NUM_THREADS   8
#define MAX_D         37        /* p < 10^37 keeps 2p inside u128 */
#define MAX_T         13
#define MAX_HITS      256
#define CKPT_MASK     0x3FFFFFF /* checkpoint every ~67.1M residues */
#define BAND_GUARD    1000      /* see docs/palindrome_split_search.md */

/* The band edge is computed EXACTLY (isqrt_u128); long double is only
 * a Newton seed, so mantissa width is not load-bearing.  An earlier
 * version estimated the edge in long double and needed an 80-bit
 * mantissa to be safe -- see docs/palindrome_split_search.md sec. 4. */

/* unsigned only for the fixed-width decimal container: no arithmetic
 * here can go negative, and we need the full 128-bit range. */
typedef unsigned __int128 u128;

static u128 pow10u[MAX_D + 2];

static void init_pow10(void) {
   pow10u[0] = 1;
   for (int i = 1; i <= MAX_D + 1; i++) {
      pow10u[i] = pow10u[i - 1] * 10;
   }
}

static inline u128 curve(int64_t n) {
   u128 v = (u128)n;
   return 2 * v * v + 2 * v + 1;
}

static inline int64_t curve_mod(int64_t r, int64_t m) {
   u128 v = (u128)r;
   return (int64_t)((2 * v * v + 2 * v + 1) % (u128)m);
}

static inline int digit_at(u128 p, int i) {
   return (int)((p / pow10u[i]) % 10);
}

static int is_pal(u128 p, int d) {
   for (int i = 0; i < d / 2; i++) {
      if (digit_at(p, i) != digit_at(p, d - 1 - i)) return 0;
   }
   return 1;
}

/* Exact integer sqrt of a u128.  Seeded from long double, then Newton,
 * then a final exact adjustment -- so the result is exact regardless of
 * how good the seed was.  Comparisons use division, never x*x, which
 * would overflow near the top of the range.  This removes floating
 * point from the band edge entirely: the band is exact, not estimated. */
static inline u128 isqrt_u128(u128 v) {
   if (v == 0) return 0;
   u128 x = (u128)sqrtl((long double)v);
   if (x == 0) x = 1;
   for (int i = 0; i < 6; i++) {
      u128 y = (x + v / x) / 2;
      if (y == x) break;
      x = y;
   }
   while (x > 1 && x > v / x) x--;
   while (x + 1 <= v / (x + 1)) x++;
   return x;
}

static inline int64_t rev_digits(int64_t v, int t) {
   int64_t h = 0;
   for (int i = 0; i < t; i++) {
      h = h * 10 + v % 10;
      v /= 10;
   }
   return h;
}

static void u128_str(u128 v, char *buf) {
   char tmp[44];
   int i = 0;
   if (v == 0) {
      strcpy(buf, "0");
      return;
   }
   while (v > 0) {
      tmp[i++] = (char)('0' + (int)(v % 10));
      v /= 10;
   }
   for (int j = 0; j < i; j++) buf[j] = tmp[i - 1 - j];
   buf[i] = '\0';
}

int main(int argc, char *argv[]) {
   if (argc < 2) {
      fprintf(stderr, "usage: %s <d> [t] [--keep5]\n", argv[0]);
      return 1;
   }

   int d = atoi(argv[1]);
   int t = 0;
   int keep5 = 0;

   for (int i = 2; i < argc; i++) {
      if (strcmp(argv[i], "--keep5") == 0) keep5 = 1;
      else t = atoi(argv[i]);
   }
   if (t == 0) t = d / 4;
   if (t < 1) t = 1;

   if (d < 1 || d > MAX_D) {
      fprintf(stderr, "d must be in [1, %d] (u128 limit)\n", MAX_D);
      return 1;
   }
   if (t > MAX_T || t >= d) {
      fprintf(stderr, "t must be in [1, min(%d, d-1)]\n", MAX_T);
      return 1;
   }

   init_pow10();
   omp_set_num_threads(NUM_THREADS);

   int64_t mod = (int64_t)pow10u[t];
   u128 step = pow10u[d - t];

   u128 hit_p[MAX_HITS];
   int64_t hit_n[MAX_HITS];
   int nhits = 0;

   printf("d=%d t=%d keep5=%d outer=10^%d threads=%d\n",
          d, t, keep5, t, NUM_THREADS);
   fflush(stdout);

   double t0 = omp_get_wtime();

#pragma omp parallel for schedule(dynamic, 4096)
   for (int64_t r = 0; r < mod; r++) {
      if ((r & CKPT_MASK) == 0 && r > 0) {
#pragma omp critical
         {
            fprintf(stderr, "  d=%2d: residue %lld / %lld\n",
                    d, (long long)r, (long long)mod);
            fflush(stderr);
         }
      }

      int64_t low = curve_mod(r, mod);        /* low t digits of p */
      if (!keep5 && low % 5 == 0) continue;   /* div-5: cannot be prime */
      if (low % 10 == 0) continue;            /* leading digit is 0 */

      int64_t high = rev_digits(low, t);      /* high t digits of p */

      /* exact: high < 10^t and step = 10^(d-t), so plo < 10^d */
      u128 plo = (u128)high * step;
      u128 phi = plo + step;
      int64_t nlo = (int64_t)((isqrt_u128(2 * plo - 1) - 1) / 2)
                    - BAND_GUARD;
      int64_t nhi = (int64_t)((isqrt_u128(2 * phi - 1) - 1) / 2)
                    + BAND_GUARD;
      if (nlo < 0) nlo = 0;

      int64_t delta = ((r - nlo) % mod + mod) % mod;

      for (int64_t n = nlo + delta; n <= nhi; n += mod) {
         u128 p = curve(n);
         if (p < pow10u[d - 1] || p >= pow10u[d]) continue;
         if (2 * t < d && digit_at(p, t) != digit_at(p, d - 1 - t))
            continue;
         if (!is_pal(p, d)) continue;
#pragma omp critical
         {
            int dup = 0;
            for (int i = 0; i < nhits; i++)
               if (hit_n[i] == n) dup = 1;
            if (!dup && nhits >= MAX_HITS) {
               fprintf(stderr, "FATAL: MAX_HITS (%d) exceeded at d=%d;"
                       " result is truncated\n", MAX_HITS, d);
               fflush(stderr);
               exit(2);
            }
            if (!dup) {
               hit_p[nhits] = p;
               hit_n[nhits] = n;
               nhits++;
            }
         }
      }
   }

   double el = omp_get_wtime() - t0;

   /* insertion sort -- nhits is tiny */
   for (int i = 1; i < nhits; i++) {
      u128 kp = hit_p[i];
      int64_t kn = hit_n[i];
      int j = i - 1;
      while (j >= 0 && hit_p[j] > kp) {
         hit_p[j + 1] = hit_p[j];
         hit_n[j + 1] = hit_n[j];
         j--;
      }
      hit_p[j + 1] = kp;
      hit_n[j + 1] = kn;
   }

   mpz_t z;
   mpz_init(z);
   for (int i = 0; i < nhits; i++) {
      char buf[44];
      u128_str(hit_p[i], buf);
      mpz_set_str(z, buf, 10);
      int pr = mpz_probab_prime_p(z, 40);
      printf("%s  n=%lld  %s\n", buf, (long long)hit_n[i],
             pr ? "PRIME" : "COMPOSITE");
   }
   mpz_clear(z);

   printf("found=%d  time=%.2fs\n", nhits, el);
   fflush(stdout);
   return 0;
}

File: docs/palsplit.py
#!/usr/bin/env python3
"""
palsplit.py -- prototype of the head/tail split search for palindromic
curve-values p = 2n^2+2n+1.

Idea under test:
  A d-digit palindrome's high t digits are the reverse of its low t
  digits.  So one choice of "n mod 10^t" fixes BOTH:
     - the low t digits of p      (directly, p = f(n) mod 10^t)
     - the high t digits of p     (= reverse of the low t digits)
  and the high digits pin n to a narrow band via n ~ sqrt(p/2).
  Intersecting "n in band" with "n = r mod 10^t" leaves ~10^(d/2-2t)
  candidates per residue, over 10^t residues.  Balanced at t ~ d/4
  this is O(10^(d/4)) instead of the O(10^(d/2)) n-enumeration.

Validation: does it recover every known palindromic curve-value?
"""
import sys, time
from math import isqrt

def f(n):
    return 2*n*n + 2*n + 1

def search(d, t):
    """All palindromic curve-values with exactly d digits."""
    M = 10**t
    lo10 = 10**(d-1)          # smallest d-digit number
    hi10 = 10**d              # one past largest
    found = []
    ops = 0
    for r in range(M):
        L = f(r) % M                       # low t digits of p
        s = str(L).zfill(t)                # "p_{t-1} ... p_0"
        if s[-1] == '0':                   # p_0 is p's LEADING digit
            continue
        H = int(s[::-1])                   # high t digits of p
        Plo = H * 10**(d-t)
        Phi = Plo + 10**(d-t)
        if Plo < lo10 or Phi > hi10:
            continue
        # n band from the magnitude of p
        nlo = (isqrt(2*Plo - 1) - 1)//2 - 2
        nhi = (isqrt(2*Phi - 1) - 1)//2 + 2
        if nlo < 0:
            nlo = 0
        # first n >= nlo with n = r (mod M)
        n = nlo + ((r - nlo) % M)
        while n <= nhi:
            ops += 1
            p = f(n)
            if Plo <= p < Phi:
                ps = str(p)
                if ps == ps[::-1]:
                    found.append((p, n))
            n += M
    return sorted(found), ops, M

def brute(d):
    """Ground-truth n-enumeration (only feasible for small d)."""
    lo10 = 10**(d-1); hi10 = 10**d
    n = (isqrt(2*lo10 - 1) - 1)//2
    while f(n) < lo10:
        n += 1
    out = []
    while True:
        p = f(n)
        if p >= hi10:
            break
        ps = str(p)
        if ps == ps[::-1]:
            out.append((p, n))
        n += 1
    return out

KNOWN = {
 13: [1635446445361, 3166046406613],
 15: [124852060258421, 149988757889941, 310433303334013],
 21: [108348382545283843801, 129052205999502250921,
      129776662212266677921, 316234169939961432613],
 25: [1427056470511150746507241, 3130000999262629990000313,
      3153012324698964232103513, 3169345937085807395439613,
      5054998070382830708994505],
 27: [108491007414868414700194801, 318216440234333432044612813,
      318288756988131889657882813],
}

if __name__ == "__main__":
    print("=== brute-force cross-check that the .txt files are complete ===")
    for d in (11, 13, 15):
        b = brute(d)
        print(f"  d={d:2d} brute: {[p for p,_ in b]}")

    print()
    print("=== head/tail split search vs known values ===")
    for d, t in ((13,3), (15,4), (21,5), (25,6), (27,7)):
        t0 = time.time()
        got, ops, M = search(d, t)
        el = time.time() - t0
        gotp = [p for p,_ in got]
        known = KNOWN[d]
        missing = [p for p in known if p not in gotp]
        extra   = [p for p in gotp if p not in known]
        status = "OK " if not missing else "MISS"
        print(f"  d={d:2d} t={t}  found={len(gotp):2d}  known={len(known)}  "
              f"outer=10^{t}  inner_ops={ops:,}  {el:6.1f}s  [{status}]")
        if missing:
            print(f"        MISSING: {missing}")
        for p in extra:
            tag = "div-5 (unlisted in file)" if p % 5 == 0 else "UNEXPECTED"
            print(f"        extra: {p}  n={dict(got)[p] if False else ''} {tag}")
        sys.stdout.flush()
